{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio import Phylo\n",
    "import pandas as pd\n",
    "import re\n",
    "from reference_based_mutation_detection import save_dict_to_pickle, find_mutations_wrt_subject_sequence\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _get_reconstructed_ancestral_sequences(ancestral_state_file_path):\n",
    "    \"\"\"\n",
    "    Reconstructs ancestral node sequences from the ancestral state file.\n",
    "    \n",
    "    Parameters:\n",
    "        - ancestral_state_file_path (str): The path to the ancestral state file.\n",
    "    \n",
    "    Returns:\n",
    "        - dict: A dictionary containing the reconstructed ancestral sequences, where the keys are the ancestral node ids and the values are the reconstructed sequences.\n",
    "    \"\"\"\n",
    "\n",
    "    # Look for the start line\n",
    "    with open(ancestral_state_file_path, 'r') as file:\n",
    "        for line_count, line in enumerate(file):\n",
    "            if line.strip().startswith(\"Node\"):\n",
    "                break  # Return the line number where the data starts\n",
    "    \n",
    "    # Read the ancestral state file\n",
    "    ancestral_states = pd.read_csv(ancestral_state_file_path, sep='\\t', skiprows=line_count)\n",
    "\n",
    "    # Concatenate the ancestral states for all the positions for a given ancestral node to reconstruct the sequence for that node\n",
    "    ancestral_sequences = {}\n",
    "    for node, group in ancestral_states.groupby('Node'):\n",
    "        sequence = ''.join(group['State'])\n",
    "        ancestral_sequences[node] = sequence\n",
    "    \n",
    "    return ancestral_sequences\n",
    "\n",
    "\n",
    "#ancestral_sequences = _get_reconstructed_ancestral_sequences(\"Sample_1.state\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import multiprocessing as mp\n",
    "import numpy as np\n",
    "\n",
    "def _process_node_group(node_group):\n",
    "    \"\"\"\n",
    "    Processes a group of rows corresponding to a single ancestral node to reconstruct its sequence.\n",
    "    \n",
    "    Parameters:\n",
    "        - node_group (pd.DataFrame): A DataFrame containing ancestral state information for a single node.\n",
    "    \n",
    "    Returns:\n",
    "        - tuple: (node_id, reconstructed_sequence)\n",
    "    \"\"\"\n",
    "    node_id = node_group['Node'].iloc[0]\n",
    "    sequence = ''.join(node_group['State'])\n",
    "    return node_id, sequence\n",
    "\n",
    "def _get_reconstructed_ancestral_sequences_parallel(ancestral_state_file_path, num_threads=1):\n",
    "    \"\"\"\n",
    "    Reconstructs ancestral node sequences from the ancestral state file using parallel processing.\n",
    "    \n",
    "    Parameters:\n",
    "        - ancestral_state_file_path (str): The path to the ancestral state file.\n",
    "    \n",
    "    Returns:\n",
    "        - dict: A dictionary containing the reconstructed ancestral sequences, where the keys are the ancestral node IDs\n",
    "          and the values are the reconstructed sequences.\n",
    "    \"\"\"\n",
    "    # Look for the start line\n",
    "    with open(ancestral_state_file_path, 'r') as file:\n",
    "        for line_count, line in enumerate(file):\n",
    "            if line.strip().startswith(\"Node\"):\n",
    "                break  # Get the line number where the data starts\n",
    "\n",
    "    # Read the ancestral state file\n",
    "    ancestral_states = pd.read_csv(ancestral_state_file_path, sep='\\t', skiprows=line_count)\n",
    "\n",
    "    # Split the data by 'Node'\n",
    "    node_groups = [group for _, group in ancestral_states.groupby('Node')]\n",
    "\n",
    "    # Use multiprocessing to process each node's sequence in parallel\n",
    "    with mp.Pool(processes=num_threads) as pool:\n",
    "        results = pool.map(_process_node_group, node_groups)\n",
    "\n",
    "    # Convert results into a dictionary\n",
    "    ancestral_sequences = dict(results)\n",
    "    \n",
    "    return ancestral_sequences\n",
    "\n",
    "\n",
    "ancestral_sequences = _get_reconstructed_ancestral_sequences_parallel(\"Sample_1.state\", 30)\n",
    "print(ancestral_sequences, 30)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _find_ancestral_node(tree, node_name):\n",
    "    \"\"\"\n",
    "    To find the nearest ancestral node to a particular node in the tree.\n",
    "\n",
    "    Parameters:\n",
    "        - tree: The phylogenetic tree to search in.\n",
    "        - node_name: The name of the node for which we wish to find the nearest ancestor.\n",
    "\n",
    "    Returns:\n",
    "        - ancestral_node: The ancestral node if found, None otherwise.\n",
    "    \"\"\"\n",
    "    # Search the target node in the tree\n",
    "    target_node = next((clade for clade in tree.find_clades() if str(clade.name).split(\"/\")[0] == node_name), None)\n",
    "    if target_node is None:\n",
    "        print(f\"Node '{node_name}' not found.\")\n",
    "        return None\n",
    "\n",
    "    # Traverse the tree to find the nearest ancestral node to the target node\n",
    "    ancestral_node = None\n",
    "    for clade in tree.find_clades():\n",
    "        if target_node in clade:\n",
    "            ancestral_node = clade\n",
    "            break\n",
    "    \n",
    "    #return nearest ancestral node\n",
    "    return ancestral_node\n",
    "\n",
    "tree = Phylo.read(\"Sample_1.treefile\", \"newick\")\n",
    "\n",
    "#ancestral_node = _find_ancestral_node(tree,\"Node1\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def _get_mutation_data_for_ancestral_nodes(ancestral_sequences, msa_file_path, tree, outgroup_id):\n",
    "    \"\"\" \n",
    "    To compute the mutations for the internal nodes of the tree.\n",
    "\n",
    "    Args:\n",
    "        ancestral_sequences (dict): Dictionary of reconstructed ancestral sequences.\n",
    "        msa_file_path (str): The path to the mutiple sequence alignment file.\n",
    "        tree (tree object): The phylogenetic tree.\n",
    "        outgroup_id(str or None): The outgroup sequence ID \n",
    "\n",
    "    Returns:\n",
    "         - denovo_mutations: A dictionary mapping each internal node to its denovo mutations.\n",
    "         - ancestral_mutations: A dictionary mapping each internal node to its ancestral mutations.\n",
    "         - ancestral_node_data: A dictionary mapping each internal node to its nearest ancestral node.\n",
    "\n",
    "    \"\"\"\n",
    "    # Initialize the dictionaries to store the mutation data\n",
    "    denovo_mutations = {}\n",
    "    ancestral_mutations = {}\n",
    "    ancestral_node_data = {}\n",
    "\n",
    "    # Sort the ancestral nodes based on their node number so that older ancestral nodes comes first. This is done to detect the ancestral mutations that are inherited. \n",
    "    ancestral_nodes = list(ancestral_sequences.keys())\n",
    "    ancestral_nodes = sorted(ancestral_nodes, key=lambda node: int(re.search(r'\\d+', node).group()))\n",
    "    \n",
    "    # Sequence records in multifasta file\n",
    "    records = list(SeqIO.parse(msa_file_path, \"fasta\"))\n",
    "\n",
    "    # Iterate through each ancestral node\n",
    "    for node in ancestral_nodes:\n",
    "        ancestral_node = _find_ancestral_node(tree, node)\n",
    "        # If it is the initial node, then the reference sequence is used as the ancestral sequence\n",
    "        if node == \"Node1\":\n",
    "            # If the outgroup_id is specified\n",
    "            if outgroup_id:\n",
    "                # Search for corresponding record in the multifasta file\n",
    "                outgroup_record = next((record for record in records if record.id == outgroup_id), None)\n",
    "                # If it is not found raise and error\n",
    "                if not outgroup_record:\n",
    "                    raise ValueError(f\"Outgroup sequence with ID '{outgroup_id}' not found in the input FASTA file.\")\n",
    "                else:\n",
    "                    # Otherwise compute the mutations\n",
    "                    outgroup_seq = str(outgroup_record.seq).upper()\n",
    "                    denovo_mutations[node] = find_mutations_wrt_subject_sequence(outgroup_seq, ancestral_sequences[node])\n",
    "                    ancestral_mutations[node] = []\n",
    "                    ancestral_node_data[\"Node1\"] = outgroup_seq    \n",
    "            else:\n",
    "                # If the outgroup_id is not specified then continue\n",
    "                continue\n",
    "        else:\n",
    "            # If it is not the initial node, then we find the nearest ancestor and compute the mutations against it\n",
    "            denovo_mutations[node] = find_mutations_wrt_subject_sequence(ancestral_sequences[str(ancestral_node).split(\"/\")[0]], ancestral_sequences[node])\n",
    "            ancestral_mutations[node] = (denovo_mutations[str(ancestral_node).split(\"/\")[0]]\n",
    "                                        + ancestral_mutations[str(ancestral_node).split(\"/\")[0]]\n",
    "                                        )\n",
    "            ancestral_node_data[node] = ancestral_node\n",
    "\n",
    "    return denovo_mutations, ancestral_mutations, ancestral_node_data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def _get_mutation_data_for_leaf_nodes(input_msa_fasta_path, ancestral_sequences, denovo_mutations, ancestral_mutations, tree, ancestral_node_data, outgroup_id):\n",
    "    \"\"\"\n",
    "    To compute the mutations for the leaf nodes of the tree\n",
    "    \n",
    "    Parameters:\n",
    "        - input_msa_fasta_path (str): Path to the multiple sequence alignment file .\n",
    "        - ancestral_sequences (dict): Dictionary of reconstructed ancestral sequences.\n",
    "        - denovo_mutations (dict): Dictionary to store denovo mutations. \n",
    "        - ancestral_mutations (dict): Dictionary to store ancestral mutations. \n",
    "        - tree: the Phylogenetic tree object.\n",
    "        - ancestral_node_data (dict): Dictionary to store ancestral node data.\n",
    "\n",
    "    Returns:\n",
    "        - denovo_mutations (dict): Updated dictionary of denovo mutations.\n",
    "        - ancestral_mutations (dict): Updated dictionary of ancestral mutations.\n",
    "        - ancestral_node_data (dict): Updated dictionary of ancestral node data for each leaf node.\n",
    "    \"\"\"\n",
    "    # Get the leaf node names from the alignment file, ignoring the outgroup sequence    \n",
    "    leaf_nodes = [record.id for record in SeqIO.parse(input_msa_fasta_path, \"fasta\") if record.id != outgroup_id]\n",
    "\n",
    "    #Compute the mutations for each leaf node\n",
    "    for leaf_node in leaf_nodes:\n",
    "        #find its nearest ancestral node data\n",
    "        ancestral_node = _find_ancestral_node(tree, leaf_node)\n",
    "        #store the ancestral node data for the leaf node\n",
    "        ancestral_node_data[leaf_node] = ancestral_node\n",
    "        #extract the sequence of this leaf node\n",
    "        sequence = next((str(record.seq).upper() for record in SeqIO.parse(input_msa_fasta_path, \"fasta\") if record.id == leaf_node), None)\n",
    "        #compute the mutations against the nearest ancestral sequence\n",
    "        denovo_mutations[leaf_node] = find_mutations_wrt_subject_sequence(\n",
    "            ancestral_sequences[str(ancestral_node).split(\"/\")[0]], \n",
    "            sequence)\n",
    "        #bootstrapping during tree construction can modify the names of the nodes in the tree by including the bootstrap results, hence we need to extract only the node name from the tree\n",
    "        #and use it to save mutation data for that leafnode\n",
    "        ancestral_mutations[leaf_node] = denovo_mutations[str(ancestral_node).split(\"/\")[0]] + ancestral_mutations[str(ancestral_node).split(\"/\")[0]]\n",
    "    return denovo_mutations, ancestral_mutations, ancestral_node_data\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def detect_mutations_phylogeny_based(input_msa_fasta_path, output_dir, tree_file_path, reconstructed_ancestral_state_file_path, outgroup_id, include_ancestral_node=False):\n",
    "    \"\"\" \n",
    "    Detect mutations in a phylogeny-based manner using an input multiple sequence alignment (MSA), \n",
    "    a phylogenetic tree, and reconstructed ancestral state sequences.\n",
    "\n",
    "    This function analyzes mutations for both internal (ancestral) and leaf nodes in a given tree \n",
    "    and outputs mutation data while considering an outgroup for rooting. If `include_ancestral_node` \n",
    "    is set to False, mutations from internal nodes are excluded.\n",
    "\n",
    "    Args:\n",
    "        input_msa_fasta_path (str or Path): Path to the input MSA FASTA file.\n",
    "        output_dir (str or Path): Directory to store the output mutation files.\n",
    "        tree_file_path (str or Path): Path to the phylogenetic tree file in Newick format.\n",
    "        reconstructed_ancestral_state_file_path (str or Path): Path to the ancestral state reconstruction file.\n",
    "        outgroup_id (str): Identifier for the outgroup sequence in the tree.\n",
    "        include_ancestral_node (bool, optional): Whether to include mutation data for ancestral nodes. Defaults to False.\n",
    "\n",
    "    Returns:\n",
    "        tuple: \n",
    "            - denovo_mutations (dict): Dictionary mapping sequences to their de novo mutations.\n",
    "            - ancestral_mutations (dict): Dictionary mapping sequences to their ancestral mutations.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Read the tree file\n",
    "    tree = Phylo.read(tree_file_path, \"newick\")\n",
    "\n",
    "    # we now reconstruct the ancestral sequences of the tree from the ancestral state file\n",
    "    print(f\"Recontructing the ancestral sequences\")\n",
    "    ancestral_sequences=_get_reconstructed_ancestral_sequences(reconstructed_ancestral_state_file_path)\n",
    "\n",
    "    # We now compute the mutations for the internal nodes of the tree\n",
    "    print(f\"Computing mutations for the ancestral nodes of the tree\")\n",
    "    denovo_mutations, ancestral_mutations,ancestral_node_data=_get_mutation_data_for_ancestral_nodes(ancestral_sequences, input_msa_fasta_path, tree, outgroup_id)\n",
    "\n",
    "    #we now compute the mutations for the leaf nodes of the tree   \n",
    "    print(f\"Computing mutations for the leaf nodes of the tree\")\n",
    "    denovo_mutations, ancestral_mutations,ancestral_node_data=_get_mutation_data_for_leaf_nodes(input_msa_fasta_path, ancestral_sequences, denovo_mutations,\n",
    "                                                                                                 ancestral_mutations, tree, ancestral_node_data, outgroup_id)\n",
    "\n",
    "    # If include_ancestral_nodes flag is set to false, we remove the ancestral nodes mutation data\n",
    "    if not include_ancestral_node:\n",
    "        denovo_mutations = {key: value for key, value in denovo_mutations.items() if not key.startswith(\"Node\")}\n",
    "        ancestral_mutations = {key: value for key, value in denovo_mutations.items() if not key.startswith(\"Node\")}\n",
    "\n",
    "    # If the outgroup sequence is specified we pop it out from the mutation data\n",
    "    if outgroup_id:\n",
    "        denovo_mutations.pop(outgroup_id)\n",
    "        ancestral_mutations.pop(outgroup_id)\n",
    "    \n",
    "    print(f\"Saving the mutation files to {output_dir}\")\n",
    "    # Save denovo mutations\n",
    "    save_dict_to_pickle(denovo_mutations, f\"{output_dir}/denovo_mutations.pkl\")\n",
    "\n",
    "    # Save ancestral mutations\n",
    "    save_dict_to_pickle(ancestral_mutations, f\"{output_dir}/ancestral_mutations.pkl\")\n",
    "\n",
    "\n",
    "    return denovo_mutations, ancestral_mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_msa_fasta_path = \"SARS-CoV-2_Sample_1_4000+1_msa.fasta\"\n",
    "output_dir = \"\"\n",
    "tree_file_path = \"Sample_1.treefile\"\n",
    "reconstructed_ancestral_state_file_path = \"Sample_1.state\"\n",
    "outgroup_id = \"NC_045512.2\"\n",
    "detect_mutations_phylogeny_based(input_msa_fasta_path, output_dir, tree_file_path, reconstructed_ancestral_state_file_path, outgroup_id, include_ancestral_node=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
