{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from itertools import product\n",
    "import pickle\n",
    "import multiprocessing\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to compute the trinucleotide frequency of a sequence to be used in normalization step while computing the trinucleotide mutation rate\n",
    "def get_kmer_frequency(sequence, k = 3):\n",
    "    \"\"\"\n",
    "    Calculate the frequency of k-mers in a given sequence.\n",
    "\n",
    "    Parameters:\n",
    "        - sequence (str): The input RNA sequence.\n",
    "        - k (int): The length of the k-mer. Default is 3 \n",
    "\n",
    "    Returns:\n",
    "        - kmers_frequency_dict: A dictionary containing all possible k-mers as keys and their frequencies in the input sequence as values.\n",
    "    \"\"\"\n",
    "    # Remove gaps from the sequence if any\n",
    "    sequence = sequence.replace(\"-\", \"\")\n",
    "    #Total number of kmers in the sequence\n",
    "    total_kmers = len(sequence) - k + 1\n",
    "\n",
    "    # Declare and Initialize the dictionary to store the kmer frequency by assigning all possible k-mers with frequency 0\n",
    "    nucleotides = [\"C\", \"T\", \"G\", \"A\"]\n",
    "    kmers_frequency_dict = {''.join(combo): 0 for combo in product(nucleotides, repeat=k)}\n",
    "\n",
    "    # Iterate through the sequence to find all k-mers\n",
    "    for i in range(total_kmers):\n",
    "        kmer = sequence[i:i + k]  # Extract the k-mer\n",
    "        # Check if the k-mer consists only of A, T, G, or C and not any other characters\n",
    "        if set(kmer).issubset(set(nucleotides)):\n",
    "            # Increment the frequency of this k-mer in the kmer_frequency dictionary\n",
    "            kmers_frequency_dict[kmer] += 1  \n",
    "    # Return the kmer dictionary\n",
    "    return kmers_frequency_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _find_mutations_wrt_subject_sequence(subject_seq, sample_seq):\n",
    "    \"\"\"\n",
    "    To compute the mutations by comparing each sequence against its nearest reconstructed ancestral sequence\n",
    "\n",
    "    Parameters:\n",
    "        - subject_seq (str): The ancestor sequence.\n",
    "        - sample_seq (str): The sample sequence.\n",
    "\n",
    "    Returns:\n",
    "        - mutations (list): A list of mutations, where each mutation is represented as a list containing the mutation, the trinucleotide context, and the trinucleotide frequency\n",
    "    \"\"\"\n",
    "    # Initialize the list to store the mutations\n",
    "    mutations = []\n",
    "    # Get the trinucleotide mutation frequencies in the ancestor sequence\n",
    "    trinucleotide_mutation_frequency = get_kmer_frequency(subject_seq, 3)\n",
    "    # Iterate through each position and bases in the two sequences\n",
    "    for i, (a, b) in enumerate(zip(subject_seq, sample_seq)):\n",
    "        # If the base is different between the ancestor and sample sequences \n",
    "        if a != b:\n",
    "            #position of the mutation in the alignment is index + 1\n",
    "            position = i + 1  \n",
    "            # Extract the trinucleotide context of the mutation\n",
    "            trinucleotide = str(subject_seq[position-2:position+1]).upper()  \n",
    "            # Base in the ancestor sequence\n",
    "            ref_base = a.upper()\n",
    "            # Base in the sample sequence\n",
    "            sample_base = b.upper()\n",
    "            # We ignore those mutations at the starting or end position of the alignment as their trinucleotide context cannot be captured\n",
    "            # We also ignore those mutations in which the surrounding trinucleotide contains abnormal basses or the mutation itself contains abnormal bases\n",
    "            if (\n",
    "                any(base not in [\"C\", \"T\", \"G\", \"A\"] for base in trinucleotide) or\n",
    "                any(base not in [\"C\", \"T\", \"G\", \"A\"] for base in [ref_base, sample_base]) or\n",
    "                len(trinucleotide) != 3\n",
    "            ):                \n",
    "                continue\n",
    "            else:\n",
    "                # Record the mutation data which includes the mutation, its trinucleotide context and trinuleotide frequency \n",
    "                mutation = f\"{ref_base}{position}{sample_base}\".upper()\n",
    "                mutations.append([mutation, trinucleotide, trinucleotide_mutation_frequency[trinucleotide]])\n",
    "               \n",
    "    return mutations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "def _detect_mutations_batchwise(batch, reference_sequence):\n",
    "    \"\"\"\n",
    "    Process a batch of sequences to detect mutations in parallel.\n",
    "    \n",
    "    Parameters:\n",
    "        - batch (list): A list of SeqRecord objects to be processed.\n",
    "        - reference_sequence (str): The reference genome sequence.\n",
    "    \n",
    "    Returns:\n",
    "        - batch_mutations_dict (dict): A dictionary mapping sequence IDs to their mutation data.\n",
    "    \"\"\"\n",
    "    # Dictionary to record mutation data for this batch\n",
    "    batch_mutations_dict = {}\n",
    "    # Iterate through each record in this batch\n",
    "    for record in batch:\n",
    "        # Extract the sample sequence and ID\n",
    "        sample_sequence = str(record.seq)\n",
    "        sample_sequence_id = record.id\n",
    "        # Compute the mutation\n",
    "        batch_mutations_dict[sample_sequence_id] = find_mutations_wrt_subject_sequence(reference_sequence, sample_sequence)\n",
    "    return batch_mutations_dict\n",
    "\n",
    "def mutation_detection_wrt_reference(aligned_file_path, ref_seq_id=None, num_threads=1):\n",
    "    \"\"\"\n",
    "    Detect mutations by comparing sequences against the reference genome using multiprocessing.\n",
    "\n",
    "    Parameters:\n",
    "        - aligned_file_path (str): The path to the MSA input file.\n",
    "        - ref_seq_id (str): The ID of the reference genome.\n",
    "        - num_threads (int): The number of processes to use.\n",
    "\n",
    "    Returns:\n",
    "        - mutations_dict (dict): A dictionary containing mutation data for each genome.\n",
    "    \"\"\"\n",
    "    # Load sequences\n",
    "    records = list(SeqIO.parse(aligned_file_path, \"fasta\"))\n",
    "\n",
    "    # Determine the reference sequence\n",
    "    if ref_seq_id:\n",
    "        reference_record = next((record for record in records if record.id == ref_seq_id), None)\n",
    "        if not reference_record:\n",
    "            raise ValueError(f\"Reference sequence with ID '{ref_seq_id}' not found in the input FASTA file.\")\n",
    "        reference_sequence = str(reference_record.seq)\n",
    "    else:\n",
    "        reference_record = records[0]\n",
    "        reference_sequence = str(reference_record.seq)\n",
    "\n",
    "    # Split records into batches for multiprocessing\n",
    "    batch_size = max(1, len(records) // num_threads)  # Avoid zero batch size\n",
    "    batches = [records[i:i + batch_size] for i in range(0, len(records), batch_size)]\n",
    "\n",
    "    # Use multiprocessing to process batches\n",
    "    with multiprocessing.Pool(processes=num_threads) as pool:\n",
    "        results = pool.starmap(_detect_mutations_batchwise, [(batch, reference_sequence) for batch in batches])\n",
    "\n",
    "    # Merge results into a final dictionary\n",
    "    mutations_dict = {}\n",
    "    for batch_result in results:\n",
    "        mutations_dict.update(batch_result)\n",
    "\n",
    "    return mutations_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def save_dict_to_pickle(dictionary, file_path):\n",
    "    \"\"\"\n",
    "        Saves a dictionary to a file using pickle.\n",
    "    \n",
    "    Parameters:\n",
    "        - dictionary(dictionary object): The dictionary to save\n",
    "        - file_path(str): The filename to save the dictionary to\n",
    "    \"\"\"\n",
    "    with open(file_path, 'wb') as file:\n",
    "        pickle.dump(dictionary, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_dict = mutation_detection_wrt_reference(\"SARS-CoV-2_Sample_1_4000+1_msa.fasta\", num_threads=32)\n",
    "#save_dict_to_pickle(mutations_dict, \"mutadict\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting typer\n",
      "  Downloading typer-0.15.1-py3-none-any.whl.metadata (15 kB)\n",
      "Requirement already satisfied: click>=8.0.0 in /home/prittam-goswami/anaconda3/lib/python3.12/site-packages (from typer) (8.1.7)\n",
      "Requirement already satisfied: typing-extensions>=3.7.4.3 in /home/prittam-goswami/anaconda3/lib/python3.12/site-packages (from typer) (4.12.2)\n",
      "Collecting shellingham>=1.3.0 (from typer)\n",
      "  Downloading shellingham-1.5.4-py2.py3-none-any.whl.metadata (3.5 kB)\n",
      "Requirement already satisfied: rich>=10.11.0 in /home/prittam-goswami/anaconda3/lib/python3.12/site-packages (from typer) (13.3.5)\n",
      "Requirement already satisfied: markdown-it-py<3.0.0,>=2.2.0 in /home/prittam-goswami/anaconda3/lib/python3.12/site-packages (from rich>=10.11.0->typer) (2.2.0)\n",
      "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /home/prittam-goswami/anaconda3/lib/python3.12/site-packages (from rich>=10.11.0->typer) (2.15.1)\n",
      "Requirement already satisfied: mdurl~=0.1 in /home/prittam-goswami/anaconda3/lib/python3.12/site-packages (from markdown-it-py<3.0.0,>=2.2.0->rich>=10.11.0->typer) (0.1.0)\n",
      "Downloading typer-0.15.1-py3-none-any.whl (44 kB)\n",
      "Downloading shellingham-1.5.4-py2.py3-none-any.whl (9.8 kB)\n",
      "Installing collected packages: shellingham, typer\n",
      "Successfully installed shellingham-1.5.4 typer-0.15.1\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m24.3.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m25.0\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install typer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
